src/database/mod.rs

```diff
@@ -158,6 +168,108 @@ pub trait Database: BatchOperations {
+                .filter_map(|out| self.del_utxo(&out.outpoint).transpose())
+                .collect::<Result<Vec<_>, _>>()?;
+            Ok(deleted_utxos)
+        }
+    }
+
+    /// Delete UTXOs based on the number of confirmations or
+    /// threshold size (number of spent utxos in DB) defined in [`DelCriteria`]
+    fn del_spent_utxos_by_criteria(
+        &mut self,
+        criteria: DelCriteria,
+        current_block_height: Option<u32>,
+    ) -> Result<Vec<LocalUtxo>, Error> {
+        let spent_utxos: Vec<LocalUtxo> = self
+            .iter_utxos()?
+            .into_iter()
+            .filter(|utxo| utxo.is_spent)
+            .collect();
+        let tx_details = spent_utxos
+            .iter()
+            .filter_map(|utxo| self.get_tx(&utxo.outpoint.txid, false).transpose())
+            .collect::<Result<Vec<_>, _>>()?;
+
+        let txs_conf_heights = tx_details
+            .iter()
+            .filter_map(|details| details.confirmation_time.as_ref().map(|conf| conf.height))
+            .collect::<Vec<_>>();
+        let utxos_heights: Vec<(&LocalUtxo, &u32)> =
+            spent_utxos.iter().zip(txs_conf_heights.iter()).collect();
+
+        let mut to_delete: Vec<&LocalUtxo> = vec![];
+        let mut failed_confirmation_criteria: Vec<(&LocalUtxo, &u32)> = vec![];
+
+        // Choose all utxos to delete by confirmation criteria.
+        if criteria.confirmations.is_some() {
+            if current_block_height.is_none() {
+                return Err(Error::Generic(String::from(
+                        "You must have a non-None `current_block_height` when using the `confirmations` criteria",
+                    )));
+            }
+
+            // split utxos_heights pair based on confirmation criteria
+            let (pass, fail): (Vec<_>, Vec<_>) =
+                utxos_heights.iter().partition(|(_utxo, height)| {
+                    (current_block_height.unwrap() - **height) >= criteria.confirmations.unwrap()
+                });
+
+            // add utxos that passed confirmation criteria test to `to_delete`
+            to_delete.extend(
+                pass.iter()
+                    .map(|(utxo, _height)| *utxo)
+                    .collect::<Vec<&LocalUtxo>>(),
+            );
+
+            // add utxos that failed the test to `failed_confirmation_criteria`
+            failed_confirmation_criteria.extend(fail.iter());
+        }
+
+        // apply threshold criteria on spent utxos
+        if criteria.threshold_size.is_some() {
+            if failed_confirmation_criteria.is_empty() {
+                failed_confirmation_criteria.extend(utxos_heights.iter());
+            }
+
+            // only select on threshold if there are spent utxos to be deleted
+            if spent_utxos.len() - to_delete.len() > 0 {
+                let qty_to_delete = spent_utxos.len() as u64
+                    - criteria.threshold_size.unwrap()
+                    - to_delete.len() as u64;
```

Comment:
I think this could underflow as well... you might want to use https://doc.rust-lang.org/std/primitive.u32.html#method.saturating_sub to avoid problems

```
let qty_to_delete = spent_utxos.len() as u64
                               - to_delete.len() as u64
                               .saturanting_sub(criteria.threshold_size.unwrap());
```

(I didn't compile this code, it probably needs some more parenthesis to work)